core/CMakeLists.txt

```diff
@@ -0,0 +1,72 @@
+set(CORE_ROOT ${CMAKE_CURRENT_SOURCE_DIR})
+
+# Appends the provided sources to the SOURCES variable of the parent.
+macro(add_sources_to_parent GENERAL_SOURCES THREADED_SOURCES UNTHREADED_SOURCES)
+    set(TEMP ${GENERAL_SOURCES})
+    if(NUMBER_OF_WORKERS)
+        list(APPEND TEMP ${THREADED_SOURCES})
+    else()
+        list(APPEND TEMP ${UNTHREADED_SOURCES})
+    endif(NUMBER_OF_WORKERS)
+    foreach(FILE ${TEMP})
+        file(RELATIVE_PATH RELATIVE_TO_ROOT ${CORE_ROOT} ${CMAKE_CURRENT_SOURCE_DIR}/${FILE})
+        list(APPEND SOURCES ${RELATIVE_TO_ROOT})
+    endforeach()
+    set(SOURCES ${SOURCES} PARENT_SCOPE)
+endmacro()
+
+# Applies the X as a preprocessor definition if X is defined.
+macro(define X)
```

Comment:
What is this for?

Reply:
The user passes arguments to CMake via the command line, like
```
cmake .. -DNUMBER_OF_WORKERS=4 -DFEDERATED=1
```

This macro passes those variables along to the compiler as preprocessor definitions. The changes I just added to the README will try to explain this.